File: src/python/problems/GoodNodes.py
# https://leetcode.com/problems/count-good-nodes-in-binary-tree/
from core.ProblemBase import *

class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class GoodNodes:
    def Solution(self, root: TreeNode) -> int:
        stack = []
        stack.append((root, float('-inf')))
        count = 0
        while stack:
            node, parentMax = stack.pop()

            count += 1 if node.val >= parentMax else 0
            parentMax = max(parentMax, node.val)

            if node.left:
                stack.append((node.left, parentMax))
            
            if node.right:
                stack.append((node.right, parentMax))

        return count
    
    def Solution1(self, root: TreeNode) -> int:
        def Dfs(node, parentMax) -> int:
            if not node:
                return 0

            count = 1 if node.val >= parentMax else 0
            parentMax = max(parentMax, node.val)
            
            count += Dfs(node.left, parentMax)
            count += Dfs(node.right, parentMax)
            return count

        return Dfs(root, float('-inf'))


if __name__ == '__main__':
    TestGen(GoodNodes) \
        .Add(lambda tc: tc.Param("nums", [1,3,4,2,2]).Result(2)) \
        .Add(lambda tc: tc.Param("nums", [3,1,3,4,2]).Result(3)) \
        .Run()

File: src/csharp/Problems/GoodNodes.cs
//https://leetcode.com/problems/count-good-nodes-in-binary-tree/

namespace LeetCode.Problems;

public sealed class GoodNodes : ProblemBase
{
    [Theory]
    [ClassData(typeof(GoodNodes))]
    public override void Test(object[] data) => base.Test(data);

    public override void AddTestCases()
        => Add(it => it.ParamTree("[3,1,4,3,null,1,5]").Result(4))
          .Add(it => it.ParamTree("[3,3,null,4,2]").Result(3))
          .Add(it => it.ParamTree("[1]").Result(1));

    private int Solution(TreeNode root)
    {
        var stack = new Stack<(TreeNode, int)>();
        stack.Push((root, int.MinValue));
        var count = 0;
        while (stack.Any())
        {
            var (node, parentMax) = stack.Pop();

            count += node.val >= parentMax ? 1 : 0;

            parentMax = Math.Max(parentMax, node.val);

            if (node.left != null)
            {
                stack.Push((node.left, parentMax));
            }

            if (node.right != null)
            {
                stack.Push((node.right, parentMax));
            }
        }

        return count;
    }

    private int Solution1(TreeNode root)
    {
        int Dfs(TreeNode? node, int parentMax)
        {
            if (node == null)
            {
                return 0;
            }

            var count = node.val >= parentMax ? 1 : 0;

            parentMax = Math.Max(parentMax, node.val);

            count += Dfs(node.left, parentMax);
            count += Dfs(node.right, parentMax);

            return count;
        }

        return Dfs(root, int.MinValue);
    }
}